dojo/utils.py

```diff
@@ -50,55 +50,151 @@
+    be used, so you should choose between product, engagement or test.
+    The lookup is done based on the deduplication_algorithm of the given finding test.
+
+    Args:
+        finding (:model:`dojo.Finding`): Finding to be matched
+        product (:model:`dojo.Product`, optional): Product to filter findings by
+        engagement (:model:`dojo.Engagement`, optional): Engagement to filter findings by
+        test (:model:`dojo.Test`, optional): Test to filter findings by
+    """
+    if product:
+        custom_filter_type = 'product'
+        custom_filter = {'test__engagement__product': product}
+
+    elif engagement:
+        custom_filter_type = 'engagement'
+        custom_filter = {'test__engagement': engagement}
+
+    elif test:
+        custom_filter_type = 'test'
+        custom_filter = {'test': test}
+
+    else:
+        raise ValueError('No product, engagement or test provided as argument.')
+
+    deduplication_algorithm = finding.test.deduplication_algorithm
+
+    deduplicationLogger.debug(
+        'Matching finding %i:%s to existing findings in %s %s using %s as deduplication algorithm.',
+        finding.id, finding.title, custom_filter_type, list(custom_filter.values())[0], deduplication_algorithm
+    )
+
+    if deduplication_algorithm == 'hash_code':
+        return (
+            Finding.objects.filter(
+                **custom_filter,
+                hash_code=finding.hash_code
+            ).exclude(hash_code=None)
+            .exclude(id=finding.id)
+            .order_by('id')
+        )
+
+    elif deduplication_algorithm == 'unique_id_from_tool':
+        return (
+            Finding.objects.filter(
+                **custom_filter,
+                unique_id_from_tool=finding.unique_id_from_tool
+            ).exclude(unique_id_from_tool=None)
+            .exclude(id=finding.id)
+            .order_by('id')
+        )
+
+    elif deduplication_algorithm == 'unique_id_from_tool_or_hash_code':
+        query = Finding.objects.filter(
+            Q(**custom_filter),
+            (
+                (Q(hash_code__isnull=False) & Q(hash_code=finding.hash_code)) |
+                (Q(unique_id_from_tool__isnull=False) & Q(unique_id_from_tool=finding.unique_id_from_tool))
+            )
+        ).exclude(id=finding.id).order_by('id')
+        deduplicationLogger.debug(query.query)
+        return query
+
+    elif deduplication_algorithm == 'legacy':
+        # This is the legacy reimport behavior. Although it's pretty flawed and
+        # doesn't match the legacy algorithm for deduplication, this is left as is for simplicity.
+        # Re-writing the legacy deduplication here would be complicated and counter-productive.
+        # If you have use cases going through this section, you're advised to create a deduplication configuration for your parser
+        logger.debug("Legacy dedup. In case of issue, you're advised to create a deduplication configuration in order not to go through this section")
```

Comment:
```suggestion
        logger.debug("Legacy dedupe. In case of issue, you're advised to create a deduplication configuration in order not to go through this section")
```